joomla/joomla-cms
# Fix for showon !: handling

Resolves issue https://github.com/joomla/joomla-cms/issues/14257
Corrects javascript coding errors in https://github.com/joomla/joomla-cms/pull/13476

Pull Request for Issue #14257 .

### Summary of Changes

There are logical errors in the javascript in linkedoptions in cms-uncompressed.js that stop the !: showon operator from working in most cases.

1. The primary problem is that `for (var i in itemval)` loops over non-enumerable properties and prototypes which means that != will ALWAYS match 
2. secondary issue is that multiple select tags get a null from `itemval = $field.val();` which causes problems.
3. radio and checkboxes return a blank value in the current code which always matches a != condition

### Testing Instructions

Create a form with this XML 
```javascript
<field name="masterfield" type="radio" default="1" class="btn-group btn-group-yesno" label="MasterField" >
	<option value="0">unknown</option>
	<option value="1">value 1</option>
	<option value="2">value 2</option>
</field>
<field name="subfield" type="radio" default="1" showon="masterfield!:0" class="btn-group btn-group-yesno" label="SubField" >
	<option value="1">Show text field</option>
	<option value="2">Show URL field</option>
</field>
<field name="subText" showon="masterfield!:0[AND]subfield:1" type="text" default="" label="Text - should appear then MasterField is !unknown and subfield is text" />
<field name="subURL" showon="masterfield!:0[AND]subfield:2" size="20" type="url" default="" label="URL - should appear then MasterField is !unknown and subfield is url" />        

<field name="selectlist" type="list" multiple="multiple" default="0" label="Multiple Select Conditions" >
	<option value="0">select a value</option>
	<option value="1">value 1</option>
	<option value="2">value 2</option>
	<option value="3">value 3</option>
	<option value="4">value 4</option>
</field>
<field name="extratext" showon="selectlist!:0" type="text" default="" label="Text should appear when something other than 'select a value' is selected" />

<field name="selectlist2" type="list" default="0" label="Single Select Conditions" >
	<option value="0">select a value</option>
	<option value="1">value 1</option>
	<option value="2">value 2</option>
	<option value="3">value 3</option>
	<option value="4">value 4</option>
</field>
<field name="extratext2" showon="selectlist2!:0" type="text" default="" label="Text should appear when something other than 'select a value' is selected" />

```

For Masterfield - choose 1 or 2 
For SubField choose text or url

Multiple Select Conditions - choose
1. no values, 
2. 'select a value' and "value 1"
3, just 'select a value'
4. just "value 1"

Single Select Conditions - choose
1.  'select a value'
2. "value 1"

### Expected result

The Text or URL fields should appear and switch as you select SubField values

Multiple Select Conditions:
1. 'text' field should be visible
2. 'text' field should be visible
3. 'text' field should NOT be visible
4. 'text' field should be visible

Single Select Conditions
1. 'text' field should NOT be visible
2. 'text' field should be visible

### Actual result

The  text and url fields  are not affected by the value of Masterfield

Multiple Select Conditions:
1. 'text' field is visible - correct
2. 'text' field is visible - correct
3. 'text' field is visible -  INCORRECT
4. 'text' field is visible - correct

Single Select Conditions
1. 'text' field is visible -  INCORRECT
2. 'text' field is visible -  correct

### Documentation Changes Required

p.s. cms-compressed.js will also need changing but I don't know how to generate this.
